### PR TITLE
Add macros for creating null-terminated string literals

### LINKED ISSUE #484: Support wide strings
Wide strings are common enough in win32 programming that it'd be a helpful to provide utilities that make it easier to work with them. In particular providing a macro for constructing wide strings at compile time. There is a fairly popular crate that provides some of these utilities already: [widestring](https://crates.io/crates/widestring), but that crate uses winapi (for obvious reasons). It's not clear what the best path forward is.

### LINKED ISSUE #581: Create PWSTR and PSTR from literals
Right now literals need to be copied into new buffers every time one is passed into a function that expects a `PWSTR`. Instead we should provide some type of macro convenience for creating `PSTR` and `PWSTR` from literals. These would encode the literals properly and add a trailing `\0` at the end all at compile time. Bonus points if we can make macro that works for either kind of string pointer.

### PR DESCRIPTION
**Pull request by kennykerr**
This update adds simple `s!` and `w!` macros that produce null terminated UTF8 and UTF16 string literals that supports both Win32 and WinRT APIs. This is quite convenient as Rust does not provide null terminated string literals nor does it provide UTF16 literals. The resulting expressions point to `const` data that is both folded by the compiler and stored in constant pages. 

Fixes #484
Fixes #581

### CODE REVIEW

File: crates/libs/windows/src/core/literals.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                            buffer[output_pos] = 0xd800 + (code_point >> 10) as u16;
+                            output_pos += 1;
+                            buffer[output_pos] = 0xdc00 + (code_point & 0x3ff) as u16;
+                            output_pos += 1;
+                        }
+                    }
+                    &{ buffer }
+                };
+                const HEADER: ::windows::core::HSTRING_HEADER = ::windows::core::HSTRING_HEADER { flags: 0x11, len: (OUTPUT_LEN - 1) as u32, padding1: 0, padding2: 0, ptr: OUTPUT.as_ptr() };
+                unsafe { ::std::mem::transmute(&HEADER) }
```
Comment (rylev) [inline crates/libs/windows/src/core/literals.rs:41]:
```suggestion
               // SAFETY: an `HSTRING` is exactly equivalent to a pointer to an `HSTRING_HEADER`
                unsafe { ::std::mem::transmute::<&::windows::core::HSTRING_HEADER, ::windows::core::HSTRING>(&HEADER) }
```

Using explicit types when using `transmute` is a safer option. This way, there's no risk of code else where in this `const` changing and then changing the type inference in unexpected ways.

**Review by rylev — COMMENTED** [review]
Looks good! I do have some nits though.

File: crates/libs/windows/src/core/literals.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test() {
+        assert_eq!(decode_utf8(b"123", 0), Some((0x31, 1)));
+        assert_eq!(decode_utf8(b"123", 1), Some((0x32, 2)));
+        assert_eq!(decode_utf8(b"123", 2), Some((0x33, 3)));
```
[the remark this answers is not in the document]
Comment (kennykerr) [reply crates/libs/windows/src/core/literals.rs:142]: My test crate disappeared. 🙄

File: crates/libs/windows/src/core/literals.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
+/// Simply terminates the literal UTF-8 string.
+#[macro_export]
+macro_rules! s {
+    ($s:literal) => {
+        ::windows::core::PCSTR(::std::concat!($s, '\0').as_ptr())
```
[the remark this answers is not in the document]
Comment (kennykerr) [reply crates/libs/windows/src/core/literals.rs:5]: Embedded nulls are sometimes allowed but usually ignored by Windows APIs.

File: crates/libs/windows/src/core/literals.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            return None;
+        }
+        return Some((result, pos));
+    }
+    None
+}
+
+#[doc(hidden)]
+#[repr(C)]
+pub struct HSTRING_HEADER {
```
[the remark this answers is not in the document]
Comment (kennykerr) [reply crates/libs/windows/src/core/literals.rs:116]: The HSTRING reference header is a subset of the HSTRING header.

File: crates/libs/windows/src/core/hstring.rs [hunk 5]
```diff
@@ -30,12 +30,7 @@ impl HSTRING {
 
     /// Get the string as 16-bit wide characters (wchars).
     pub fn as_wide(&self) -> &[u16] {
-        if self.is_empty() {
-            return &[];
-        }
-
-        let header = self.0;
-        unsafe { core::slice::from_raw_parts((*header).data, (*header).len as usize) }
+        unsafe { core::slice::from_raw_parts(self.as_ptr(), self.len()) }
```
Comment (rylev) [inline crates/libs/windows/src/core/hstring.rs:33]: This is incorrect. From the docs:

> data must be non-null and aligned even for zero-length slices.

**Review by rylev — DISMISSED** [review]

**Pull request merged**